staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go

```diff
@@ -356,3 +368,61 @@ func tooManyRequests(req *http.Request, w http.ResponseWriter, retryAfter string
 	w.Header().Set("Retry-After", retryAfter)
 	http.Error(w, "Too many requests, please try again later.", http.StatusTooManyRequests)
 }
+
+func newRequestWaitContextFactory(clock utilsclock.PassiveClock, defaultRequestWaitLimit time.Duration) *requestWaitContextFactory {
+	return &requestWaitContextFactory{
+		clock:                   clock,
+		defaultRequestWaitLimit: defaultRequestWaitLimit,
+		newCtxFn:                context.WithDeadline,
+	}
+}
+
+type requestWaitContextFactory struct {
+	clock utilsclock.PassiveClock
+
+	// the default wait duration that is used if the request context does
+	// not have any deadline, the wait limited used for:
+	// (a) initialization of a watch or
+	// (b) a request whose context has no deadline
+	defaultRequestWaitLimit time.Duration
+
+	// comes handy to write unit test
+	newCtxFn func(parent context.Context, at time.Time) (context.Context, context.CancelFunc)
+}
+
+// GetRequestWaitContext returns a new context with a deadline of how
+// long the request is allowed to wait before it is removed from its
+// queue and rejected.
+// The context.CancelFunc returned must never be nil and the caller is
+// responsible for calling the CancelFunc function for cleanup.
+//   - ctx: the context associated with the request (it may or may
+//     not have a deadline).
+//   - isWatch: true if the given request is a WATCH, otherwise false
+func (rwcf *requestWaitContextFactory) GetRequestWaitContext(ctx context.Context, isWatch bool) (context.Context, context.CancelFunc) {
+	if ctx.Err() != nil {
+		return ctx, func() {}
+	}
+
+	reqArrivedAt := rwcf.clock.Now()
+	if reqReceivedTimestamp, ok := apirequest.ReceivedTimestampFrom(ctx); ok {
+		reqArrivedAt = reqReceivedTimestamp
+	}
+
+	if isWatch {
+		return rwcf.newCtxFn(ctx, reqArrivedAt.Add(rwcf.defaultRequestWaitLimit))
+	}
+
+	thisReqWaitLimit := rwcf.defaultRequestWaitLimit
+	if deadline, ok := ctx.Deadline(); ok {
+		// a) we will allow the request to wait in the queue for one
+		// fourth of the time of its allotted deadline.
+		// b) if the request context does not have any deadline
+		// then we default to 'defaultRequestWaitLimit'
+		thisReqWaitLimit = deadline.Sub(reqArrivedAt) / 4
+		if thisReqWaitLimit <= 0 {
+			return ctx, func() {}
+		}
```

Comment:
What would go wrong if these lines were deleted?

Reply:
+1

Reply:
it would create a new context with a deadline that has already passed, but, since we check for `ctx.Err()` at the beginning of the function the chance of it happening is very low indeed.  (i removed the lines)

https://github.com/golang/go/blob/70e04706d8ccd4c93de54e1f5c7b15c942018dee/src/context/context.go#L631-L634